### DISCUSSION TITLE
Make Categories = Structures?

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cms#9781] [Ideas] [answered]
There could bits I am missing, but it seems that Categories are just Structures with less features.

- no revisions/drafts
- no "types" (EntryType, CategoryType)
- no authors

I've found the difference can be hard for clients to delineate, which is understandable.

I'm wondering if something should be done to make this clearer. e.g:

- Give Categories the same features as Structures (just with a different name/place)
- Remove categories in favor of Structures 🙉

**Comment by brandonkelly** [comment]
The main difference is in how they get related to entries. When you relate a nested Structure section entry, that’s the only thing that gets related. When you related a nested category, its parents get related as well.

But yeah, I’m sure there are cases where it would make sense for categories to have authors; there are cases where entries _shouldn’t_ have authors. Etc etc. For now we’ve just given each element type the features that they most commonly have in other systems.

**Comment by timkelty** [comment]
Thanks for the explanation - Auto-relating parents for categories is definitely something I wasn't considering.

> The main difference is in how they get related to entries

Just spitballing - Since what you're describing is more a characteristic of the Entires/Categories _field_, if one were looking for more parity between Structures & Categories, a field setting of "Relate structural parents?" might work as well.

> there are cases where entries shouldn’t have authors

[Indeed. Run in to that before](https://github.com/craftcms/cms/issues/896#issuecomment-287080149). In the future, perhaps both categories and sections could have the options:

- Enable versioning for entries in this section?
- Allow entries to have authors?

**Comment by brandonkelly** [comment]
Yeah realistically I’d rather just merge the two concepts, rather than duplicate a bunch of entry features for categories and vise-versa. But I wouldn’t want to merge the concepts until we can come up with a way to customize how the "entries" / "categories" can be displayed for content managers (e.g. #1662).

**Comment by timkelty** [comment]
Thanks - subscribing to that thread.

**Comment by carlcs** [comment]
I wanted to propose the same thing, so thanks @timkelty for the issue. We’re facing the dilemma of deciding whether to go with categories or entries on every Craft install. Currently we have to make the decision based on the way the associated relations field type works, or if a field layout select (entry types) is ever needed, and we usually end up with elements used for classification in both places, obviously bad for content managers.

@brandonkelly don’t you think customizable sources view in the entries index is good enough to implement this request before https://github.com/craftcms/cms/issues/1662, which is probably still a good time away (considered for 4.0).

**Comment by brandonkelly** [comment]
@carlcs Realistically we’re not going to drop a major concept like Categories outside of a major release cycle (and 3.0 GA is way too close for us to consider it for this one). That said, if there’s anything you’d like to see us do to make using Structured entries a more viable replacement for categories in the meantime (e.g. allow Entries fields to be more like Categories fields), feel free to post those suggestions as separate issues.

**Comment by carlcs** [comment]
@brandonkelly understandable. And a good idea with making the Entries field type more advanced as a first step. Thanks

**Comment by timkelty** [comment]
> And a good idea with making the Entries field type more advanced as a first step.

Yep - if we wanted it as a 3.0 feature, I'd say the biggest thing would be an option to  "Include ancestors when related".

Only question then is - is that an option when creating a Structure section, or just a field option for Entries fields. I'm guessing the latter would be most flexible, and maybe easier to implement as well.

**Comment by brandonkelly** [comment]
Yeah seems like the obvious place to put it is in the Entries field settings.

**Comment by timkelty** [comment]
@brandonkelly FWIW, here's what my plan has been:

### Make a new "Enhanced Elements" field type

This would be a catch-all replacement for all applicable element types, with additional field options to address the ancestor selection issue, as well as several other enhancements:

- **First field settings option would be "Entry Type"**
  - Presumably, there would be a way for devs to register an element type for this field, which would make it show up as an option. This would be great for plugin devs, as it would allow them to easily have a field for their element type without having to roll their own.
- **Next would be the "Display/Selection" type:**
  - Options include: Default (for selected element type), Element Index, Tag List, Checkbox/Radios/Dropdown (like https://github.com/timkelty/craft-simplecategories)
    - What would be cool about this is you could have Entries, but with a Tag List/Creation UI.
    - A simplified element selection is a frequent request from clients, specifically with Categories, where they aren't usually creating new ones. They really want to just bang through entries and use checkboxes, not the element index modal.
- **Next would be Element type specific options:**
  - The normal options for that entry type: eg "Limit/Branch Limit", "Sources"
  - Add "Include ancestors on selection" and "Branch Limit" for Entries

While it is on my "someday" list to implement this as a plugin, I think having the ability for plugin devs to automatically register their entry types is a good reason this should be implemented in core. For example, Field types like [Freeform Form](http://docs.solspace.com/craft/freeform/v2/overview/fieldtype.html#creating-a-freeform-form-field) would no longer be necessary.

Another cool concept would be that a plugin dev could register a new element "view type" UI, that would become available when installed. E.g. - if the above were implemented w/o a "Checkboxes" view type, I could write a "Simple Element View Types" plugin, instead of having to make various field types: "SimpleCategories", "SimpleElements", etc.

**Comment by JeanLucEsser** [comment]
Hi there,

I see this looks to be on the 5.0 milestone, which I understand as it is a major core concept to be changed.

But... are you planning on adding an "include ancestors on selection" option to the structures field (on a 3.x or 4.x release) before replacing categories entirely in 5.0 so we can use them as a full replacement to categories and benefit from their added power (multisite I'm looking at you)? In core or via plugin. Or maybe I missed a plugin that exists already?

Just a quick update on this to know where you stand would allow us to adapt accordingly. Thanx ;)

**Comment by carlcs** [comment]
@JeanLucEsser both issues (this one and https://github.com/craftcms/cms/issues/2632) are on the same milestone list right now. It looks like Entries field updates won’t make it into a Craft 3 release in the end.

**Comment by JeanLucEsser** [comment]
@carlcs yes that's what I thought. So in the meantime I guess I'll use structures because I need multisite ability, but I'll add a query to find all descendant ids of the structure parent's entry before querying the content related to all those ids. Guess it's the best way to go for now. 

The real downside is having no visual cue of the ancestors tree for the selected entry in the structure field. Which by the way I find it to be an issue even when using structures as intended with no relation to ancestors, as it could be useful information. But it's tricky to implement, as each entry should be independent and able to be re ordered... maybe on hover or something... anyway, not the subject at hand, I digress.

Thanx anyway for getting back to me. Much appreciated.

**Comment by carlcs** [comment]
I agree with what you said about the label of entries in the entries field. In structures the title isn’t always enough info for an entry to be recognizable.

**Comment by brandonkelly** [comment]
Sorry this is actually intended to be part of 4.0. Just fixed that.

**Comment by iuscare** [comment]
I am wondering if this Is this still valid for 4.0, because categories still exist in Craft 4 beta 1.

**Reply by timkelty** [reply]
@iuscare it is. Categories now have drafts, but the selection field itself is still different, in that it automatically selects ancestors.

We may add an option to add this behavior to entries fields in a 4.x release, which would allow entries to be a replacement for categories.

**Reply by davidhellmann** [reply]
Sounds good. What I also miss in Categories is Propagation methods. I think its also fine that categories have a future but the features should nearly the same as in structures. 

Split entries and categories brings a bit more structure maybe. Important is to know how it works in the future. We use just entries cause the statement that categories have no future 😅

**Reply by thomassausen** [reply]
So, what's the recommended way to add Categories or Tags to your articles? 

Because of this discussion, I changed all categories to structures to future-proof it. Just to recognize that Categories will be there for some time.

**Reply by davidhellmann** [reply]
I'll stay with structures. I don't use tags since five years. And we moved cats to structures because more features and I like the approach to have "content" instead of "entries" which was the initial plan for this I think.

**Comment by myleshyson** [comment]
> @iuscare it is. Categories now have drafts, but the selection field itself is still different, in that it automatically selects ancestors.


@timkelty Hey just following up to see if there's any concrete plans yet to add category like relationship logic to entry fields? Currently we're stuck using categories because it's very important for our site that various elements are related up the tree, and it's helpful that the category fields _shows_ the relationship up the tree. However it's concerning for our content team that categories don't support revisions like entries.

**Reply by timkelty** [reply]
@myleshyson Plans, yes. It is just a matter of when we can prioritize it.
If it helps, it is something that is planned for the 4.x release cycle.

Once we have a good idea of when we can fit it in, I'll make sure to post here.

In the meantime, if anyone is inclined, PRs are certainly welcome. I have one started here (thought pretty stale at this point) : https://github.com/craftcms/cms/pull/11088

**Comment by brandonkelly** [comment, marked as answer]
We have decided to move forward with this. Categories, tags, and global sets are all going to become entries.

Here’s how it will play out:

**Craft 4.4**

- Entries fields are going to gain new settings that enable them to behave like Tags and Entries fields. ([#11088](https://github.com/craftcms/cms/pull/11088), [#11379](https://github.com/craftcms/cms/discussions/11379))
- A new CLI command will be added which will aid in migrating existing categories, tags, and global sets over to entries.
- Categories, tags, and global sets will be formally deprecated/discouraged going forward.

**Craft 5**

- The UI for managing categories and global sets will be removed.
- The category, tag, and global set element types will remain, so developers can choose to migrate to entries _after_ the Craft 4 → 5 upgrade if they wish, rather than forcing it as a prerequisite to upgrading to Craft 5.

**Craft 6**

- Category, tag, and global set element types will be fully removed.

**Reply by JeanLucEsser** [reply]
This is top news! 

Just a friendly reminder about the need for structures to have a sense of context visually about its tree (obvious for categories as the tree is part of the selection but sorely needed for structures where the selected item may not be enough to get a sense of the tree it belongs to). This is a UX challenge in some ways, maybe we can discuss it in its own thread or wait for the 4.4 beta cycle.

Anyway, thank you @brandonkelly for moving forward on this, seems like tags as entries could make them manageable as well, we'll see what you have in mind, can't wait to play with 4.4 beta (should you make one) ;)

**Reply by brandonkelly** [reply]
> Just a friendly reminder about the need for structures to have a sense of context visually about its tree.

Sounds like something that would be addressed independently from the rest of this stuff. Mind creating a new discussion about it with more details?

**Reply by JeanLucEsser** [reply]
Sure: https://github.com/craftcms/cms/discussions/12196

**Reply by Mosnar** [reply]
> **Craft 4.4**
> 
> * Entries fields are going to gain new settings that enable them to behave like Tags and Entries fields. ([#11088](https://github.com/craftcms/cms/pull/11088), [#11379](https://github.com/craftcms/cms/discussions/11379))
> * A new CLI command will be added which will aid in migrating existing categories, tags, and global sets over to entries.
> * Categories, tags, and global sets will be formally deprecated/discouraged going forward.

With Global Sets being deprecated, what will the recommended approach be going forward? Would it just become a Single?

**Reply by RyanRoberts** [reply]
Liking the sound of this but I'm curious to see how this looks in the CP, at the moment having a Settings single within "Entries" feels odd.

**Reply by nixondesigndev** [reply]
Will things such as the "Categories" field for selecting "Matches items" within sales in Commerce be changed to support entries?

**Reply by thupsi** [reply]
> Liking the sound of this but I'm curious to see how this looks in the CP, at the moment having a Settings single within "Entries" feels odd.

I suppose this will be something that the upcoming "Unified content view" will attempt to solve, among other things?

**Reply by brandonkelly** [reply]
> I suppose this will be something that the upcoming "Unified content view" will attempt to solve, among other things?

Correct

**Reply by kellyannmcnamara01** [reply]
@brandonkelly, I have a couple of follow-up questions that I have no doubt have simple and well-thought-out answers but I am just trying to anticipate the workback adjustments for a site currently being built. 

1. Will the CLI command also adjust the fields in the editor (while keeping the handle) and maintain the relationships?
2. In `crafts.enties()` queries, will we still be using the `relatedTo()` param, but instead of passing the category/tag we pass the single?
3. Is there documentation/examples we can refer to, in order to best prepare?

I look forward to this new path, I think it makes a lot of sense. Any insight you can provide would be greatly appreciated. We are trying to launch in April, so it would be grand to have the code adjusted and prepared even if 4.4 is coming later.

**Reply by brandonkelly** [reply]
@kellyannmcnamara01 

1. Yes, Categories fields would be converted to Entries fields with the proper settings to maintain existing behavior.
2. Not 100% sure I’m following you there (no categories would be converted to a Single section/entry), but yes, in general your `relatedTo()` params wouldn’t need to change.
3. No immediate action will need to be taken when 4.4 comes out, but certainly we will have docs explaining everything by the time 5.0 is out.

**Reply by brandonkelly** [reply]
4.4 is out now! There’s also a new blog post that goes into a little more background and detail about the planning of these changes: https://craftcms.com/blog/entrification

**Comment by Mosnar** [comment]
We discussed this internally and had some concerns about how stuffing categories in the Entries sidebar along with the other sections is going to impact the author experience. We have a number of sites with upwards of 20 category groups. I'm all for converting them to sections/entries; however, scrolling  through tons of sections is going to be tough to convince clients is an improvement. One idea might be to convert categories to entries "behind the scenes" until the new unified content view is ready. Perhaps this could be facilitated with a checkbox or toggle in the section settings. As long as this was enabled, these sections would remain fenced off as categories-in-name-only, but they'd use the same templates and logic as entries.

**Reply by timkelty** [reply]
@Mosnar Worth noting that in 4.x, this would be completely optional. Categories would remain as-is, but Entries would gain an option to enforce hierarchy. A CLI utility would let you convert a group to a section, but nothing would happen automatically.

**Reply by Mosnar** [reply]
What do you think the deciding factor should be on whether we build new websites using categories or entries? If we think we might need more than 3 category groups, stick with categories, otherwise use entries?

**Reply by timkelty** [reply]
More importantly, choose entries if you specifically need features that categories currently don't have (eg multiple field layouts, post/expiry dates).

**Reply by wsydney76** [reply]
Ignored categories from the very beginning, feels way more natural to handle everything as entries, having the freedom to group sections under meaningful headings in the entries element index, and resolving hierarchy at runtime.

And yes, that can make the sidebar long and a bit confusing, but adding some [JavaScript/CSS](https://github.com/wsydney76/craft4-ddev-theme/blob/main/modules/resources/cp/dist/cp_scripts.js) making the headers collapsible in a persistent way helped a lot.

(As long as the markup doesn't change....)

/Mel
